### ISSUE TITLE
About cogdl's reproducibility

### ISSUE THREAD

**Issue by 4fee8fea** [THUDM/cogdl#341]
As mentioned in an [issue](https://github.com/pyg-team/pytorch_geometric/issues/92) of PyG, 
> I evaluated this problem and see perfect reproducibility on CPU when using the seed_everything code from you. However, on GPU, we can not guarantee determinism because we make heavy use of scatter ops to implement Graph Neural Networks [which are non-deterministic by nature on GPU](https://pytorch.org/docs/stable/notes/randomness.html).

PyG's reproducibility is limited by the presence of the `torch_scatter` operation. 

Will the reproducibility of `cogdl` be better?

I think this is an important question because it is closely related to the ablation analysis.

**Comment by cenyk1230** [comment]
Hi @4fee8fea,
Thanks for your attention. 
We try our best to (re-)implement GNN models in our library to reproduce similar results to the original results. 
However, all GNN libraries (e.g., PyG, DGL, CogDL) cannot provide the exact reproducibility since the scatter operators on GPU are not guaranteed with determinism by nature. In reality, we can run the same experiments multiple times with different seeds (e.g., 10 seeds) to obtain a relatively accurate metric, which is enough for comparison and ablation analysis.

**Comment by 4fee8fea** [comment]
Hi @cenyk1230,

Thanks for your prompt reply and advice! 

I noticed that taking ten different random seeds to report the average is also a recommended practice in OGB. 

My experiments use the fixed seed (e.g., taking seed as 42) and run ten times within each experiment. Perhaps this is equivalent to the practice above. 

However, the average results were also different each time under this setting. 

I see now that it comes down to torch_scatter, which is somewhat frustrating. 

Thanks for the reply, though, and all the best to you!
